cpp/tests/streaming/test_shuffler.cpp

```diff
@@ -195,48 +223,70 @@ Node shuffler_nb(
         std::vector<rapidsmpf::shuffler::PartID> finished(total_num_partitions);
         std::iota(finished.begin(), finished.end(), 0);
         shuffler_ctx->shuffler->insert_finished(std::move(finished));
+
         co_return;
     };
 
     // extract task: extract the packed chunks from the shuffler and send them to the
     // output channel
-    auto extract_task = [](auto shuffler_ctx, auto ctx, auto ch_out) -> Node {
-        ShutdownAtExit c{
-            ch_out
-        };  // TODO: could this be problematic with multiple consumers?
+    auto extract_task =
+        [](auto shuffler_ctx, auto ctx, auto ch_out, auto& latch) -> Node {
         co_await ctx->executor()->schedule();
 
         while (!shuffler_ctx->shuffler->finished() || !shuffler_ctx->ready_pids.empty()) {
-            auto expected = co_await shuffler_ctx->ready_pids.pop();
-            if (!expected.has_value()) {  // queue is shutdown, so exit the loop
-                break;
+            // try pop the partition id from the queue
+            // Note: using pop() may cause a deadlock, because pop tasks could be in front
+            // of the executor queue. try_pop() gives better control over the execution
+            // order.
+            auto result = shuffler_ctx->ready_pids.try_pop();
+            rapidsmpf::shuffler::PartID pid;
+            if (result.has_value()) {
+                pid = *result;  // partition id is found
+            } else if (result.error() == coro::queue_consume_result::stopped) {
+                break;  // queue is shutdown, so exit the loop
+            } else {  // queue is empty or busy, so yield and continue
+                co_await ctx->executor()->yield();
+                continue;
             }
 
-            auto packed_chunks = shuffler_ctx->shuffler->extract(*expected);
+            auto packed_chunks = shuffler_ctx->shuffler->extract(pid);
 
             co_await ch_out->send(
-                std::make_unique<PartitionVectorChunk>(
-                    *expected, std::move(packed_chunks)
-                )
+                std::make_unique<PartitionVectorChunk>(pid, std::move(packed_chunks))
             );
 
             if (shuffler_ctx->shuffler->finished()) {
                 // if the shuffler is finished, shutdown & drain the ready_pids queue
                 co_await shuffler_ctx->ready_pids.shutdown_drain(ctx->executor());
```

Comment:
The person _producing_ values into `ready_pids` should shut down the queue I think.

Reply:
Ah ok, that's too complicated.